### PR TITLE
Fix the non-default constructor mechanism of bytecode recording

### LINKED ISSUE #31153: Bug in Quarkus bytecode generation
### Describe the bug

Hi all, I'm Ronald, the creator of [JobRunr](https://github.com/jobrunr/jobrunr), a Quartz replacement.

JobRunr also integrates with Quarkus and to do so, I created a custom Quarkus Extension. I'm currently struggling with this extension.

In this extension, I've written a [RecurringJobsFinder](https://github.com/jobrunr/jobrunr/blob/bugfix/689/framework-support/jobrunr-quarkus-extension/deployment/src/main/java/org/jobrunr/quarkus/extension/deployment/RecurringJobsFinder.java) that searches for all methods annotated with @Recurring and schedules them using a recorder so that the bytecode is recorded.

This is all tested by means of integration tests here:
https://github.com/jobrunr/jobrunr/blob/bugfix/689/framework-support/jobrunr-quarkus-extension/tests/src/main/java/org/jobrunr/quarkus/it/TestService.java#L20-L25

All of this mostly works fine but for one of these recurring jobs, I receive the following exception:

```java
java.lang.RuntimeException: Failed to start quarkus
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
    at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:625)
    at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:696)
    at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
       ...
Caused by: java.lang.RuntimeException: Failed to start quarkus
    at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
    at io.quarkus.runtime.Application.start(Application.java:101)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:253)
    at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:250)
    at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:592)
    at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:640)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$12(ClassBasedTestDescriptor.java:395)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:395)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:211)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:84)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
    ... 52 more
Caused by: java.lang.UnsupportedOperationException
    at java.base/java.util.Collections$UnmodifiableCollection.add(Collections.java:1060)
    at io.quarkus.deployment.steps.JobRunrExtensionProcessor$findRecurringJobAnnotationsAndScheduleThem1351984528.deploy_0(Unknown Source)
    at io.quarkus.deployment.steps.JobRunrExtensionProcessor$findRecurringJobAnnotationsAndScheduleThem1351984528.deploy(Unknown Source)
    ... 68 more
```

The bytecode generated by the `Recorder` is as follows (thanks @gastaldi!):
```java
    public void deploy_0(StartupContext var1, Object[] var2) {
        JobRunrRecurringJobRecorder var4 = new JobRunrRecurringJobRecorder();
        ArrayList var3 = new ArrayList();
        JobDetails var6 = new JobDetails("org.jobrunr.quarkus.it.TestService", (String)null, "aRecurringJob", (List)var3);
        Object var5 = var1.getValue("proxykey93");
        var4.schedule((BeanContainer)var5, "my-recurring-job", var6, "*/15 * * * *", (String)null, (String)null);
        ArrayList var7 = new ArrayList(1);
        JobParameter var8 = new JobParameter("org.jobrunr.jobs.context.JobContext", "org.jobrunr.jobs.context.JobContext", (Object)null);
        ((Collection)var7).add(var8);
        JobDetails var10 = new JobDetails("org.jobrunr.quarkus.it.TestService", (String)null, "anotherRecurringJob", (List)var7);
        ((Collection)var10.getJobParameters()).add(var8);
        Object var9 = var1.getValue("proxykey93");
        var4.schedule((BeanContainer)var9, "another-recurring-job-with-jobContext", var10, (String)null, "PT10M", (String)null);
    }
```

We noticed that `var8` is added twice for some reason. As `((Collection)var10.getJobParameters())` returns an unmodifiable list, the second add fails. But, I do not understand the reason why the parameter is added twice in the first place as that will also cause problems later on in JobRunr (2 parameters where the method only expects 1).

### Expected behavior

The parameter should only be added once.

### Actual behavior

The parameter is added twice.

### How to Reproduce?

1. git clone https://github.com/jobrunr/jobrunr
2. git checkout `bugfix/689`
3. import project in Intellij
4. Run test `./framework-support/jobrunr-quarkus-extension/tests/src/test/java/org/jobrunr/quarkus/it/JobRunrFunctionalityTest.java`

### Output of `uname -a` or `ver`

Darwin Ronalds-MacBook-Pro.local 21.6.0 Darwin Kernel Version 21.6.0: Thu Sep 29 20:13:56 PDT 2022; root:xnu-8020.240.7~1/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment (build 17.0.1+12-39) OpenJDK 64-Bit Server VM (build 17.0.1+12-39, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.15.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5.1

### Additional information

See also the analysis of @gastaldi at https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/JobRunr.20Quarkus.20extension.20-.20UnsupportedOperationException

### PR DESCRIPTION
**Pull request by gastaldi**
- Fix the non-default constructor mechanism of bytecode recording
- Fixes #31153
- Optimize bytecode recording test performance

### CODE REVIEW

**Comment by gastaldi** [comment]
Thanks to @Ladicek and @geoand who helped with the investigation!

**Comment by geoand** [comment]
I didn't do anything, you guys did all the work :)

**Review by geoand — APPROVED** [review]

**Comment by gastaldi** [comment]
@geoand well, you approved this PR, so that's something 😉

**Review by Ladicek — APPROVED** [review]

**Comment by gastaldi** [comment]
@Ladicek looks like ArC didn't like that change:

```
Could not find parameters for constructor public io.smallrye.config.ConfigMappings$ConfigClassWithPrefix(java.lang.Class,java.lang.String) could not read field values [arg1, arg0]
```

File: core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java [hunk 1]
```diff
@@ -1173,6 +1173,13 @@ public void prepare(MethodContext context) {
                 nonDefaultConstructorHandles[i] = loadObjectInstance(obj, existing,
                         parameterTypes[count++], relaxedValidation);
             }
+            if (nonDefaultConstructorHolder.constructor.getParameterCount() > 0) {
+                Parameter[] parameters = nonDefaultConstructorHolder.constructor.getParameters();
+                for (int i = 0; i < parameters.length; ++i) {
+                    String name = parameters[i].getName();
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java:1179]: I think we need to check if `java.lang.reflect.Parameter#isNamePresent` here, WDYT @Ladicek ?

Comment (Ladicek) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java:1179]: I was wondering why the existing code (which I merely copied) doesn't check if parameter names exist, but it seemed to work... :shrug:

Yeah, I agree we shouldn't put the synthetic parameter names `argN` to the map. That, together with how the map is used (matching parameters are removed from it, and then the map is checked if some parameters remain), should fix the issue here.

Comment (gastaldi) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java:1179]: @Ladicek mind pushing the fix to this PR?

Comment (Ladicek) [reply core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java]: Pushed a commit, let's see how that goes :-)

**Review by stuartwdouglas — APPROVED** [review]

**Comment by gastaldi** [comment]
Test failures are unrelated and happen in current `main` branch

**Comment by gsmet** [comment]
If we want it backported, we will need specific PRs. It doesn't apply cleanly and I prefer having the author fix the conflicts given it's a sensitive matter.

**Comment by gastaldi** [comment]
I'll refrain from backporting this to 2.13 because it involves changes that would require a deeper analysis (like the introduction of `RecordingAnnotationsUtil`)

**Pull request merged**